### ISSUE TITLE
insert/update with typescript generics. Syntax?

### ISSUE THREAD

**Issue by abcd-ca** [knex/knex#4772]
I would love some help or an example in the docs of how to accomplish the following use of insert/update with generics:

```
const createItem = async <T>(knex: Knex, tableName: DBTableNameEnum, item: T): Promise<T> => {
  const [createdItem] = await knex<T>(tableName).insert(item).returning('*')
  return createdItem
}
```

The complier error I get for the insert arg is, 
`TS2345: Argument of type 'T' is not assignable to parameter of type 'T extends CompositeTableType  > ? ResolveTableType  | readonly ResolveTableType [] : DbRecordArr...> | readonly DbRecordArr...>[]'.`

Moreover, I think the problem is the `T` where I have,
`await knex<T>(tableName).insert(item).returning('*')`

**Comment by ggallovalle** [comment]
My biggest suggestion is that you never try to make an "orm" or "query builder" like thing is never really type safe. Because you would be getting into a lot of assumptions and implementation details of the thing when what you are really after is something more of like the following "a function that stores an item into something and return that something that something being inserted with all the default things, like ids or so on"

```typescript
const createItem2 = <T>(into: string, item: T): Promise<T> => {
  return Promise.resolve(0 as any);
};
```

But if you insist this is a more "type" safe way, just do not put the generic `T` into any of the knex stuff.

```typescript
const createItem = async <T>(knex: Knex, tableName: DBTableNameEnum, item: T): Promise<T> => {
  const [createdItem] = await knex(tableName).insert(item).returning('*')
  return createdItem
}
```

**Comment by abcd-ca** [comment]
@ggallovalle I'm not trying to make it capable of inserting anything or making an ORM, I just have an app that happens to have a dozen or so entities that are almost all the same so I was trying to make it more DRY. Thank you for your input, I think this is good enough if I can't do `Knex<T>...`

**Comment by ggallovalle** [comment]
Yes, because what you get out of being type safe is being correct and sure about what you do without actually running the code. And with an external tool such as a query builder you can get to that, but at the end of the day it just too much work and it just cover the simplest of cases, what if you want to select just X amount of columns, or put an alias. There are so many edge cases that to me and my team is not even worth it, so we abstract the actual query builder AKA knex in this case behind a function that does makes sense for our use case, as the `createItem2`.

This is just my input and glad if it helps you, and if not happy to hear your thoughts.

**Comment by mercmobily** [comment]
This issue has been added to the TypeScript meta-issue #6248, which tracks all TypeScript typing improvements needed in Knex. It will be addressed as part of the broader effort to improve TypeScript support, particularly for insert/update operations with naming convention transformations.

The current workaround remains using `postProcessResponse` and `wrapIdentifier` hooks (see [docs](https://knexjs.org/guide/#postprocessresponse)) or the [`knex-stringcase`](https://github.com/Kequc/knex-stringcase) library.

**Comment by mercmobily** [comment]
I am closing it, but it will be still done as part of the typescript meta issue
